src/Spice86.Core/Emulator/CPU/CfgCpu/ExecutionContextManager.cs

```diff
@@ -1,61 +1,113 @@
 ﻿namespace Spice86.Core.Emulator.CPU.CfgCpu;
 
-using Spice86.Core.Emulator.CPU.CfgCpu.ControlFlowGraph;
+using Serilog.Events;
+
 using Spice86.Core.Emulator.CPU.CfgCpu.Feeder;
 using Spice86.Core.Emulator.CPU.CfgCpu.Linker;
 using Spice86.Core.Emulator.CPU.CfgCpu.ParsedInstruction;
-using Spice86.Core.Emulator.VM;
-using Spice86.Core.Emulator.VM.Breakpoint;
+using Spice86.Core.Emulator.Function;
+using Spice86.Core.Emulator.Memory;
 using Spice86.Shared.Emulator.Memory;
+using Spice86.Shared.Interfaces;
 
 public class ExecutionContextManager : InstructionReplacer {
-    private readonly EmulatorBreakpointsManager _emulatorBreakpointsManager;
-    private readonly Dictionary<SegmentedAddress, ISet<ICfgNode>> _executionContextEntryPoints = new();
+    private readonly ILoggerService _loggerService;
     private readonly CfgNodeFeeder _cfgNodeFeeder;
+    private readonly IMemory _memory;
+    private readonly State _state;
     private int _currentDepth;
+    private readonly FunctionCatalogue _functionCatalogue;
+    private readonly ExecutionContextReturns _executionContextReturns = new();
 
-    public ExecutionContextManager(EmulatorBreakpointsManager emulatorBreakpointsManager, CfgNodeFeeder cfgNodeFeeder,
-        InstructionReplacerRegistry replacerRegistry) : base(replacerRegistry) {
-        _emulatorBreakpointsManager = emulatorBreakpointsManager;
+    public ExecutionContextManager(IMemory memory,
+        State state,
+        CfgNodeFeeder cfgNodeFeeder,
+        InstructionReplacerRegistry replacerRegistry,
+        FunctionCatalogue functionCatalogue,
+        ILoggerService loggerService) : base(replacerRegistry) {
+        _loggerService = loggerService;
         _cfgNodeFeeder = cfgNodeFeeder;
-        // Initial context at init
-        CurrentExecutionContext = new(_currentDepth);
+        _memory = memory;
+        _state = state;
+        _functionCatalogue = functionCatalogue;
+        // Initial fake but non-null context at init
+        CurrentExecutionContext = NewExecutionContext(SegmentedAddress.ZERO);
     }
 
+    private int CurrentDepth {
+        get => _currentDepth;
+        set {
+            _currentDepth = value;
+            _loggerService.LoggerPropertyBag.ContextIndex = value;
```

Comment:
Is it possible to do `_loggerPropertyBag.ContextIndex = value;` instead ?

Reply:
As discussed, we already depend on _loggerService for logging in this class so let's not add an additional property